base/applications/sdbinst/sdbinst.c

```diff
@@ -0,0 +1,957 @@
+#define APPCOMPAT_LAYERS_REG_PATH L"Software\\Microsoft\\Windows NT\\CurrentVersion\\AppCompatFlags\\Layers"
+#define APPCOMPAT_INSTALLEDSDB_REG_PATH L"Software\\Microsoft\\Windows NT\\CurrentVersion\\AppCompatFlags\\InstalledSDB"
+#define UNINSTALL_REG_PATH L"Software\\Microsoft\\Windows\\CurrentVersion\\Uninstall"
+#define DBPATH_KEY_NAME L"DatabasePath"
+
+HRESULT
+RegisterSdbEntry(
+    _In_ PWCHAR sdbEntryName,
+    _In_ LPCWSTR dbGuid,
+    _In_ ULONGLONG time,
+    _In_ BOOL isInstall,
+    _In_ BOOL isExe)
+{
+    PWCHAR regName;
+    HKEY  hKey = NULL;
+    LSTATUS status;
+    HRESULT hres;
+
+    regName = (PWCHAR)HeapAlloc(GetProcessHeap(), 0, MAX_PATH * sizeof(WCHAR));
+    if (!regName)
+    {
+        hres = HRESULT_FROM_WIN32(ERROR_NOT_ENOUGH_MEMORY);
+        goto end;
+    }
+    ZeroMemory(regName, MAX_PATH * sizeof(WCHAR));
+
+    hres = StringCchPrintf(regName, MAX_PATH, L"%ls\\%ls",
+                     isExe ? APPCOMPAT_CUSTOM_REG_PATH : APPCOMPAT_LAYERS_REG_PATH, sdbEntryName);
+    if (FAILED(hres))
+    {
+        wprintf(L"StringCchPrintfW error: 0x%08X\n", hres);
+        goto end;
+    }
+
+    // Remove key
+    if (!isInstall)
+    {
+        status = RegDeleteKey(HKEY_LOCAL_MACHINE, regName);
+        return HRESULT_FROM_WIN32(status);
+    }
+
+    // Create main key
+    status = RegCreateKeyEx(HKEY_LOCAL_MACHINE,
+                            regName,
+                            0,
+                            NULL,
+                            REG_OPTION_NON_VOLATILE,
+                            KEY_ALL_ACCESS,
+                            NULL,
+                            &hKey,
+                            NULL);
+
+    if (status != ERROR_SUCCESS)
+    {
+        wprintf(L"RegKeyCreateEx error: 0x%08X", status);
+        hres = HRESULT_FROM_WIN32(status);
+        goto end;
+    }
+
+    // Set instlled time
+    status = RegSetValueEx(hKey,
+                           dbGuid,
+                           0,
+                           REG_QWORD,
+                           (PBYTE)&time,
+                           sizeof(time));
+    if (status != ERROR_SUCCESS)
+    {
+        wprintf(L"RegSetValueEx error: 0x%08X", status);
+        hres = HRESULT_FROM_WIN32(status);
+        goto end;
+    }
+
+end:
+    if (hKey)
+    {
+        RegCloseKey(hKey);
+    }
+
+    if (regName)
+    {
+        HeapFree(GetProcessHeap(), 0, regName);
+    }
+
+    return hres;
+}
+
+HRESULT
+AddUninstallKey(
+    _In_ LPCWSTR dbName,
+    _In_ LPCWSTR sdbInstalledPath,
+    _In_ LPCWSTR guidDbStr)
+{
+    PWCHAR sdbinstPath;
+    PWCHAR regName;
+    PWCHAR uninstString;
+    HKEY hKey = NULL;
+    HRESULT hres;
+
+    sdbinstPath = (PWCHAR)HeapAlloc(GetProcessHeap(), 0, MAX_PATH * sizeof(WCHAR));
+    regName = (PWCHAR)HeapAlloc(GetProcessHeap(), 0, MAX_PATH * sizeof(WCHAR));
+    uninstString = (PWCHAR)HeapAlloc(GetProcessHeap(), 0, MAX_PATH * sizeof(WCHAR));
```

Comment:
Same comment as above if you know that these buffers are of fixed size and won't be resized dynamically in the code.